fix: set conferencing apps as default

Amit91848 opened the pull request:
## What does this PR do?


- Fixes #15375 
- Fixes CAL-3903 


https://github.com/calcom/cal.com/assets/74371312/7c5658fb-ce02-4517-a088-1b5d099a0a83



## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

Amit91848 committed:
fix: set conferencing apps as default

Amit91848 commented:
> when you sign up for a new account and choose Google Meet as the default meeting platform, for some reason it still keeps Cal Video as the default and you have to go to settings to change it manually

It was installing normally and not setting it as default. I have added it to google meet only for now? Do want me to add it to all conferencing apps?

SomayChauhan commented:
Thank you, @Amit91848.

>It was installing normally and not setting it as default. I have added it to google meet only for now? Do want me to add it to all conferencing apps?

can you add it for all the oAuth-based conferencing apps

Amit91848 commented:
Sure @SomayChauhan

Amit91848 committed:
fix in other oauth conferencing apps

Amit91848 commented:
Added @SomayChauhan

ibex088 committed:
Merge branch 'main' into fix/set_conferencing_app_default_onboarding

PeerRich reviewed (dismissed):
looks nice, but can we sort video conferencing apps by popularity as well? We sort calendars already by popularity

ideally Google Meet (and Zoom etc.) is at the top

Amit91848 commented:
Will do @PeerRich

Amit91848 commented:
> looks nice, but can we sort video conferencing apps by popularity as well? We sort calendars already by popularity
> 
> ideally Google Meet (and Zoom etc.) is at the top

Just checked it is already being sorted by popularity, local development doesn't have app count so isn't being reflected @PeerRich

PeerRich commented:
ah very nice

ibex088 committed:
creating default event-types should set the user's default location

ibex088 reviewed (commented):
LGTM!
ThankYou for the fix @Amit91848 
added a small fix in https://github.com/calcom/cal.com/pull/15376/commits/20fdb50081ae684723a5af78436068c0b09509a4
where the default event-types we create for the new user were not picking up the default app we just set.

PS: For OAuth apps, where the link is generated automatically we set the newly added conferencing app as the default location for all event types. For link-based apps, users can manually set the location.

ibex088 committed:
Merge branch 'main' into pr/15376

ibex088 committed:
Update getDefaultLocations.ts

ibex088 committed:
fix: unit test

hariombalhara commented on packages/app-store/components.tsx:
```diff
@@ -31,7 +31,7 @@ export const InstallAppButtonWithoutPlanCheck = (
           disabled: props.disableInstall,
           onClick: () => {
-            mutation.mutate({ type: props.type });
+            mutation.mutate({ type: props.type, defaultInstall: true });
```
This seems to be making any app that is being installed to be come the default.

So, if I install Google Meet, it becomes default and then if I install Zoom, it can then become default replacing Google Meet.

I think the intention is make an app default only through onboarding and this `InstallButton` component is used throughout the app.

hariombalhara replied:
No. It is used inside InstallAppButton as well which is used elsewhere.

hariombalhara reviewed (changes requested):
@SomayChauhan Could you confirm the behaviour that you were expecting here?

During the onboarding, I have the option to install as many conferencing apps I want.

If I install Google Meet, it becomes default. If I then install Zoom, should it then replace Google Meet as the default?


For me what makes sense is that, if it is the very first conferencing app that you install, make it the default. Other conferencing app installations after that shouldn't be considered default. 

But I really think the best would be to make it explicit like through a CTA along with the installed app to make it default. Otherwise keep Cal Video as the default as that is the integration we heavily focus on maintaining.

SomayChauhan commented:
> But I really think the best would be to make it explicit like through a CTA along with the installed app to make it default. Otherwise keep Cal Video as the default as that is the integration we heavily focus on maintaining.

I like this Idea better 👍 , this would make a lot more sense

Amit91848 commented:
So installation will be as is, after install you want to do something similar to this? @SomayChauhan @hariombalhara 

![Screenshot 2024-07-04 190639](https://github.com/calcom/cal.com/assets/74371312/13dea035-5f09-439f-ad11-da6c9d3e44e3)

SomayChauhan commented:
![Screenshot 2024-07-05 at 10 32 18 AM](https://github.com/calcom/cal.com/assets/32706411/4dc734c7-9235-4273-8dea-3006c43550df)

I believe @hariombalhara suggested adding a `set as default` button next to the `connect` button. The following conditions should be met for this button to be visible:

1. The app should be installed.
2. The button should only be visible for OAuth apps (those that automatically generate a meeting link, such as Google Meet, Zoom, etc.) and not for link-based apps like Campfire and Discord.

Amit91848 commented:
> I believe @hariombalhara suggested adding a `set as default` button next to the `connect` button. The following conditions should be met for this button to be visible:
> 
> 1. The app should be installed.
> 2. The button should only be visible for OAuth apps (those that automatically generate a meeting link, such as Google Meet, Zoom, etc.) and not for link-based apps like Campfire and Discord.
> 

Sounds good will make changes 👍

hariombalhara commented:
@Amit91848 We would need confirmation from @ciaranha @PeerRich on it. So, could you show us how it will look first before implementing it.

Amit91848 commented:
![Screenshot 2024-07-08 133247](https://github.com/calcom/cal.com/assets/74371312/b64c4ae0-91d3-4777-a10b-8798cabde06d)

Should look somewhat like this. @hariombalhara @PeerRich @ciaranha

ciaranha commented:
> For me what makes sense is that, if it is the very first conferencing app that you install, make it the default. Other conferencing app installations after that shouldn't be considered default.

Following from @hariombalhara's point above, it's enough to make the first conferencing app you install the default and to add a badge alongside, as we already do on the Installed Apps page. The functionality of changing defaults can be done later. We don't want to over and necessarily complicate onboarding, it's supposed to be quick and simple.

Amit91848 commented:
> > For me what makes sense is that, if it is the very first conferencing app that you install, make it the default. Other conferencing app installations after that shouldn't be considered default.
> 
> Following from @hariombalhara's point above, it's enough to make the first conferencing app you install the default and to add a badge alongside, as we already do on the Installed Apps page. The functionality of changing defaults can be done later. We don't want to overcomplicate onboarding; it's supposed to be quick and simple.

Makes sense, will make changes.

Amit91848 committed:
default install only for first oauth app

Amit91848 committed:
fix: type-check

Amit91848 commented:
@hariombalhara @ciaranha made changes based on requirements, please review again.


https://github.com/calcom/cal.com/assets/74371312/2e6a8602-3d98-40c6-b09f-e8736ad9d71e

ibex088 committed:
Merge branch 'main' into fix/set_conferencing_app_default_onboarding

Amit91848 committed:
Merge branch 'main' into fix/set_conferencing_app_default_onboarding

ibex088 committed:
Merge branch 'main' into pr/15376

hariombalhara commented on apps/web/components/getting-started/steps-views/ConnectedVideoStep.tsx:
```diff
@@ -36,9 +41,13 @@ const ConnectedVideoStep = (props: ConnectedAppStepProps) => {
                       logo={item.logo}
                       installed={item.userCredentialIds.length > 0}
+                      defaultInstall={
+                        (item.isOAuth || item.slug === "google-meet") && !hasAnyInstalledVideoApps
```
I don't like checking the type of an app here. We should really check for the feature of the app.

I believe the apps that we want to support here to be set as default are those that have `isOAuth=true` i.e. they have been setup using a personal connection. But Google Meet doesn't have `isOAuth` set to true, because it is setup automatically with Google Calendar. Daily Video too doesn't have isOAuth set to true.
<img width="573" alt="Screenshot 2024-07-09 at 8 46 36 AM" src="https://github.com/calcom/cal.com/assets/1780212/e68bad68-256c-4b30-916c-64e0e3f70bff">

I think the right check here could be `metadata.appData.location.linkType==='dynamic'`. `dynamic` means the link is per booking and not a static link i.e. common for all bookings.

Amit91848 replied:
Yes this would make more sense, based on earlier conversation I assumed we were only adding it for oauth based apps.

hariombalhara commented on packages/lib/server/getDefaultLocations.test.ts:
```diff
@@ -45,7 +45,7 @@ describe("getDefaultLocation ", async () => {
     await addAppsToDb([TestData.apps["google-meet"]]);
     const res = await getDefaultLocations(user);
-    expect(res[0]).toEqual({
+    expect(res[0]).toContain({
```
What is the reason for this test to change?

hariombalhara commented on packages/lib/server/getDefaultLocations.ts:
```diff
@@ -22,7 +22,8 @@ export async function getDefaultLocations(user: User): Promise<EventTypeLocation
     const locationType = foundApp?.locationOption?.value ?? DailyLocationType; // Default to Daily if no location type is found
-    return [{ type: locationType, link: defaultConferencingData.appLink }];
+    const credentialId = foundApp.credential.id ?? undefined;
+    return [{ type: locationType, link: defaultConferencingData.appLink, credentialId }];
```
I don't really see changes in this PR that use this newly added `credentialId` prop.

Amit91848 replied:
@SomayChauhan made those changes in https://github.com/calcom/cal.com/pull/15376#pullrequestreview-2107355807

> added a small fix in https://github.com/calcom/cal.com/commit/20fdb50081ae684723a5af78436068c0b09509a4
where the default event-types we create for the new user were not picking up the default app we just set.

SomayChauhan replied:
![334563703-b13e0b85-517f-42cc-8839-75cad8422823](https://github.com/calcom/cal.com/assets/32706411/534cf40a-0364-47ea-a0ec-0ff2bafd003b)
the reason I added `credentialId` was that in some cases the Location dropdown remained empty
although I cant replicate it now, so we can remove it too

hariombalhara replied:
Yeah, let's remove that and maybe fix in a separate PR once we are able to replicate it.

hariombalhara commented on packages/app-store/office365video/api/callback.ts:
```diff
@@ -104,6 +105,10 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
 
   await createOAuthAppCredential({ appId: "msteams", type: "office365_video" }, responseBody, req);
 
+  if (state?.defaultInstall) {
+    setDefaultConferencingApp(userId, "msteams");
+  }
```
I want to avoid requiring `defaultInstall` code here per app and also requiring all new dynamic location conferencing apps to have this code too. 
It is important so that if someone adds a new conferencing app, it can be set as default during onboarding without requiring any changes in the code.

Possible Solutions:
1. Maybe there is some common function that is called for all these apps and it would make sense to move this code there. 
2. Or we simply initiate a `setDefaultConferencingApp` call from browser during onboarding after the first conferencing app is installed. 


I find Option-2 better as this default conferencing app requirement is really a requirement of onboarding only.

Amit91848 replied:
> Possible Solutions:
> 
> 1. Maybe there is some common function that is called for all these apps and it would make sense to move this code there.
> 2. Or we simply initiate a `setDefaultConferencingApp` call from browser during onboarding after the first conferencing app is installed successfully. Could be done in the `onSuccess` callback
> 
> I find Option-2 better as this default conferencing app requirement is really a requirement of onboarding only.

Cannot call `setDefaultConferencingApp` directly as it has prisma. Would need to create an endpoint for it. Will check if they have some common function

hariombalhara reviewed (changes requested):
@Amit91848 I really want to avoid making changes in the _getAdd or callback.ts of apps because that would mean a lot of apps would need to be modified and all new conferencing apps would have to be ensured to have that code

Suggested a different approach, let us see if it works

Amit91848 committed:
update default on success

Amit91848 commented:
@hariombalhara updated the code. I haven't removed default install from _add.ts and callback.ts yet, but they aren't being used. Just wanted to confirm the flow first. Couldn't find any common file for all, decided to create an endpoint to update the default because we cannot directly run `setDefaultConferencingApp` in browser.

SomayChauhan commented:
https://github.com/calcom/cal.com/pull/15702 should be merged before this pr is merged

hariombalhara commented:
The flow looks good now. @Amit91848 you can now cleanup the app files.

There is some more code improvements that are required. I would leave the comments inline though.

hariombalhara commented on packages/app-store/components.tsx:
```diff
@@ -13,15 +13,27 @@ import type { RouterOutputs } from "@calcom/trpc/react";
-  const mutation = useAddAppMutation(null);
+  const mutation = useAddAppMutation(null, {
+    onSuccess: () => {
+      if (me && props.defaultInstall && props.slug) {
+        setDefaultConferencingApp.mutate({ slug: props.slug });
+      }
```
This change makes InstallAppButton do too many things. Following the Open/Closed principle, I think it would be much better to introduce an `onSuccess` prop to the component and then let the onboarding component provide it's own onSuccess callback that would take care of the default installation.

hariombalhara replied:
We can then remove `defaultInstall` prop and `slug` prop

hariombalhara reviewed (changes requested):
Left inline comments.

Amit91848 committed:
fix: pass callback prop

Amit91848 committed:
chore: code cleanup

Amit91848 committed:
Merge branch 'main' into fix/set_conferencing_app_default_onboarding

hariombalhara commented on apps/web/components/getting-started/steps-views/ConnectedVideoStep.tsx:
```diff
@@ -36,9 +41,14 @@ const ConnectedVideoStep = (props: ConnectedAppStepProps) => {
+                      slug={item.slug}
                       installed={item.userCredentialIds.length > 0}
+                      defaultInstall={
+                        item.appData?.location?.linkType === "dynamic" && !hasAnyInstalledVideoApps
```
We need to consider only `dynamic` linkType apps here. Otherwise if someone installs, say AroundVideo first and then installs Zoom, zoom won't be considered `default` because  `hasAnyInstalledVideoApps` would be true at that time.

hariombalhara committed:
Fix condition

hariombalhara commented on apps/web/components/getting-started/steps-views/ConnectedVideoStep.tsx:
```diff
@@ -36,9 +42,14 @@ const ConnectedVideoStep = (props: ConnectedAppStepProps) => {
+                      slug={item.slug}
                       installed={item.userCredentialIds.length > 0}
+                      defaultInstall={
+                        !defaultConferencingApp && item.appData?.location?.linkType === "dynamic"
```
We should allow making a conferencing app as default only if there is no existing default conferencing app.

hariombalhara commented:
Looks good to me now !! Thanks for bearing with me.


Not approving the PR yet as another PR needs to be merged first.

ibex088 committed:
no longer needed

- as we reverted https://github.com/calcom/cal.com/commit/20fdb50081ae684723a5af78436068c0b09509a4

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into fix/set_conferencing_app_default_onboarding
